 ---
title: "Bioconductor"
output: learnr::tutorial
runtime: shiny_prerendered
---

```{r setup, include=FALSE}
```

### Install

You can install the `Bioconductor` package straight from CRAN. Try running the code below in your own RStudio environment. If you are asked whether to install all, some, or no dependencies, type "a" for all. 

If you are asked if you want to update packages, say "Yes". If a package that R tried to update can't be updated, take note of that package's name and install it separately from CRAN without downloading and compiling the more recent version. 

This might take a few minutes to run.

```{r install-Bioconductor}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install()
BiocManager::install(c("GenomicFeatures", "AnnotationDbi"))
library(BiocManager)
library()
```

Now let's see how many packages are available in `BiocManager`.

```{r available-packages}
length(BiocManager::available())
```

## Get some Data

Install more packages. This code will work only if you have the latest R version installed.

```{r prep-to-downlod-data}
if (!requireNamespace("SingleCellExperiment", quietly = TRUE)) {
    BiocManager::install("SingleCellExperiment")
}
library(SingleCellExperiment)

if (!requireNamespace("scRNAseq", quietly = TRUE)) {
    BiocManager::install("scRNAseq")
}
library(scRNAseq)

if (!requireNamespace("SummarizedExperiment", quietly = TRUE)) {
    BiocManager::install("SummarizedExperiment")
}
library(SummarizedExperiment)

if (!requireNamespace("ExperimentHub", quietly = TRUE)) {
    BiocManager::install("ExperimentHub")
}
library(SummarizedExperiment)

if (!requireNamespace("doParallel", quietly = TRUE)) {
    install.packages("doParallel")
}
library(doParallel)

if (!requireNamespace("gam", quietly = TRUE)) {
    install.packages("gam")
}
library(gam)

if (!requireNamespace("clusterExperiment", quietly = TRUE)) {
      BiocManager::install("clusterExperiment")
}

if (!requireNamespace("scone", quietly = TRUE)) {
      BiocManager::install("scone")
}

if (!requireNamespace("zinbwave", quietly = TRUE)) {
      BiocManager::install("zinbwave")
}

if (!requireNamespace("slingshot", quietly = TRUE)) {
      BiocManager::install("slingshot")
}
```

Obtain the data.

```{r obtain-data}
# Bioconductor
library(BiocParallel)
library(clusterExperiment)
library(scone)
library(zinbwave)

# GitHub
library(slingshot)

# CRAN
library(doParallel)
library(gam)
library(RColorBrewer)

set.seed(20)
```

In the next code chunk, R checks to see if you have a data directory, and if you don't, it creates it and stores its location in `data_dir`. Then it saves important URLs in `urls`.

```{r pre-processing-data}
data_dir <- "/Users/braunsb/Documents/Arcus Learning Plan Generator/Lessons/data/" 
if (!dir.exists(data_dir)) system(sprintf('mkdir %s', data_dir))

urls = c("https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE95601&format=file&file=GSE95601%5FoeHBCdiff%5FCufflinks%5FeSet%2ERda%2Egz",
         "https://raw.githubusercontent.com/rufletch/p63-HBC-diff/master/ref/oeHBCdiff_clusterLabels.txt")
```

The next code chunk contains some very good coding practice. Before downloading data from a public repository, the code checks first to see if you already have it. If not, it downloads it.

```{r check-for-file-before-downloading}
if(!file.exists(paste0(data_dir, "GSE95601_oeHBCdiff_Cufflinks_eSet.Rda"))) {
  if (!dir.exists(data_dir)) system(sprintf('mkdir %s', data_dir))
  download.file(urls[1], paste0(data_dir, "GSE95601_oeHBCdiff_Cufflinks_eSet.Rda.gz"))
  R.utils::gunzip(paste0(data_dir, "GSE95601_oeHBCdiff_Cufflinks_eSet.Rda.gz"))
  # assayData(Cufflinks_eSet)$exprs = NULL
  assayData(Cufflinks_eSet)$fpkm_table = NULL
  assayData(Cufflinks_eSet)$tpm_table = NULL
  save(Cufflinks_eSet, file='data/GSE95601_oeHBCdiff_Cufflinks_eSet_reduced.Rda')
}

if(!file.exists(paste0(data_dir, "oeHBCdiff_clusterLabels.txt"))) {
  download.file(urls[2], paste0(data_dir, "oeHBCdiff_clusterLabels.txt"))
}
```

We load the data into `Cufflinks_eSet`, establish a count matrix, then remove undetected genes from the data.

```{r load-data}
load(paste0("/Users/braunsb/Documents/Arcus Learning Plan Generator/data/GSE95601_oeHBCdiff_Cufflinks_eSet.Rda"))

# Count matrix
E <- assayData(Cufflinks_eSet)$counts_table

# Remove undetected genes
E <- na.omit(E)
E <- E[rowSums(E)>0,]
dim(E)
```

At this point, I think we've gotten you past any code in which what you do and what the website instructions say to do are different. You can continue by going to [this page on the bioconductor website](https://www.bioconductor.org/help/course-materials/2017/BioC2017/Day2/Workshops/singleCell/doc/workshop.html#pre-processing) and searching for "We remove the ERCC spike-in sequences and the CreER gene". Use the instructions there to do this workflow. 

Please enter your new code into code chunks below this point, and let me know if you have any questions. 

```{r your-new-code-here}

```

```{r more-of-your-new-code-here}

```

```{r etc}

```
